### PR TITLE
make destinationPolicy optional

### PR DESCRIPTION
**Pull request by waifa**
Updated handlers to not require destinationPolicy, aws-logs-destination.json to not list as required and added unit tests

These changes were made to resolve a github issue created by an external customer.
Customer requested destinationPolicy to not be required when creating logs destination as it is not required when creating destination via API.
Change updates the CreateHandler to not attempt action on destinationPolicy when it is not provided.
Unit tests, integration tests, and canary tests were also created to test the new edge case.

*Issue #, if available:*
AWS::Logs::Destination - DestinationPolicy should not be required #101

*Description of changes:*
As described in commit message

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-logs-destination/src/test/java/software/amazon/logs/destination/CreateHandlerTest.java [hunk 1]
```diff
@@ -283,5 +286,172 @@ public void handleRequest_Should_ReturnFailureProgressEvent_When_DestinationRead
...
+                .thenReturn(putDestinationResponse);
+
+        final ProgressEvent<ResourceModel, CallbackContext> response =
+                handler.handleRequest(proxy, request, new CallbackContext(), proxyClient, logger);
+
+        assertThat(response).isNotNull();
+        assertThat(response.getStatus()).isEqualTo(OperationStatus.SUCCESS);
+        assertThat(response.getCallbackDelaySeconds()).isEqualTo(0);
+
+        //following line causes error, response doesn't include value for dest.policy it seems, but desiredResourceState does include it
```
Comment (evanuk) [inline aws-logs-destination/src/test/java/software/amazon/logs/destination/CreateHandlerTest.java:322]: Is this an old comment?

File: aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java [hunk 2]
```diff
@@ -182,4 +182,45 @@ public void handleRequest_Should_ReturnSuccess_When_DestinationIsFound() {
         assertThat(response.getErrorCode()).isNull();
     }
 
+    // tests for optional parameter destination policy not provided tests
+    @Test
+    public void handleRequest_Should_ReturnSuccess_When_DestinationIsFound_and_DestinationPolicyNotProvided() {
+        final Destination destination_wihtout_policy = getTestDestination(false);
```
Comment (evanuk) [inline aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java:188]: nit: typo and should be camelCase

File: aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java [hunk 3]
```diff
@@ -182,4 +182,45 @@ public void handleRequest_Should_ReturnSuccess_When_DestinationIsFound() {
         assertThat(response.getErrorCode()).isNull();
     }
 
+    // tests for optional parameter destination policy not provided tests
+    @Test
+    public void handleRequest_Should_ReturnSuccess_When_DestinationIsFound_and_DestinationPolicyNotProvided() {
```
Comment (evanuk) [inline aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java:187]: Can we add a test for DestinationIsFound, DestinationPolicyNotProvided, and the initial destination had a destination policy?

File: aws-logs-destination/src/main/java/software/amazon/logs/destination/CreateHandler.java [hunk 4]
```diff
@@ -36,8 +37,8 @@
                 }))
                 .then(progress -> putDestination(proxy, callbackContext, proxyClient, model, DESTINATION_CREATE_GRAPH,
                         logger, Action.CREATE))
-                .then(progress -> putDestinationPolicy(proxy, callbackContext, proxyClient, model,
-                        DESTINATION_POLICY_CREATE_GRAPH, logger, Action.CREATE))
+                .then(progress -> model.getDestinationPolicy()!=null ? putDestinationPolicy(proxy, callbackContext, proxyClient, model,
```
[the remark this answers is not in the document]
Comment (waifa) [reply aws-logs-destination/src/main/java/software/amazon/logs/destination/CreateHandler.java:40]: I thought it was more readable without spacing, but updated

File: aws-logs-destination/src/test/java/software/amazon/logs/destination/CreateHandlerTest.java [hunk 5]
```diff
@@ -48,6 +48,8 @@
 
     private Destination destination;
 
+    private Destination destination_without_policy;
```
[the remark this answers is not in the document]
Comment (waifa) [reply aws-logs-destination/src/test/java/software/amazon/logs/destination/CreateHandlerTest.java:51]: good catch

File: aws-logs-destination/src/test/java/software/amazon/logs/destination/CreateHandlerTest.java (hunk 1 above)
Comment (waifa) [reply aws-logs-destination/src/test/java/software/amazon/logs/destination/CreateHandlerTest.java:322]: oops, yep added that during testing

File: aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java (hunk 2 above)
Comment (waifa) [reply aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java:188]: not sure why VS Code didn't point that one out, thanks!

Comment (waifa) [reply aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java:188]: oh I see, its a different file, I defined it with the typo

File: aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java (hunk 3 above)
Comment (waifa) [reply aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java:187]: Hmm, "initial destination had a destination policy" implying the resource is created with the destination policy but the update doesn't have a destination policy? If that is what you mean, that case is in the integration testing. AFAIU, unit tests only test one function, so none of the tests in these files call more than one function.

Comment (waifa) [reply aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java:187]: Never mind, I see what you mean, added test case

File: aws-logs-destination/src/test/java/software/amazon/logs/destination/CreateHandlerTest.java [hunk 6]
```diff
@@ -283,5 +286,172 @@ public void handleRequest_Should_ReturnFailureProgressEvent_When_DestinationRead
...
+
+        assertThat(response).isNotNull();
+        assertThat(response.getStatus()).isEqualTo(OperationStatus.FAILED);
+        assertThat(response.getErrorCode()).isEqualTo(HandlerErrorCode.AlreadyExists);
+    }
+
+    @Test
+    public void handleRequest_Should_ThrowCfnResourceConflictException_When_PutOperationIsAborted_and_DestinationPolicyNotProvided() {
+        final DescribeDestinationsResponse describeResponse = DescribeDestinationsResponse.builder()
+                .destinations(destination_without_policy)
```
Comment (wchunl) [inline aws-logs-destination/src/test/java/software/amazon/logs/destination/CreateHandlerTest.java]: do we care about whether or not DescribeDestination returns a destination with a policy? it seems like we throw "ResourceNotFoundException" on line 440 anyways.

File: aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java (hunk 3 above)
Comment (wchunl) [reply aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java:187]: was this supposed to be DestinationNotFound? if the DestinationIsFound doesnt that mean the resource already exist and should fail?

File: aws-logs-destination/src/main/java/software/amazon/logs/destination/CreateHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (wchunl) [reply aws-logs-destination/src/main/java/software/amazon/logs/destination/CreateHandler.java]: just wondering but if the customer provides a bad destination policy (i.e. the format is invalid) will this throw?

File: aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java (hunk 3 above)
Comment (waifa) [reply aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java]: This is the handler for update. I think you are confusing this with create. Create fails if the destination already exists. Update should fail in the case that destination does not exist, and should not fail when the destination does exist(given  of course there aren't any other unrelated errors).

File: aws-logs-destination/src/main/java/software/amazon/logs/destination/CreateHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (waifa) [reply aws-logs-destination/src/main/java/software/amazon/logs/destination/CreateHandler.java]: Yes PutDestinationPolicy method in BaseHandlerStd makes the API call and handles the exception of a failed create

File: aws-logs-destination/src/test/java/software/amazon/logs/destination/CreateHandlerTest.java (hunk 6 above)
Comment (waifa) [reply aws-logs-destination/src/test/java/software/amazon/logs/destination/CreateHandlerTest.java]: agreed

File: aws-logs-destination/src/test/java/software/amazon/logs/destination/CreateHandlerTest.java [hunk 7]
```diff
@@ -283,5 +286,172 @@ public void handleRequest_Should_ReturnFailureProgressEvent_When_DestinationRead
...
+
+        assertThat(response.getResourceModel()).isEqualTo(request.getDesiredResourceState());
+        assertThat(response.getResourceModels()).isNull();
+        assertThat(response.getMessage()).isNull();
+        assertThat(response.getErrorCode()).isNull();
+    }
+
+    @Test
+    public void handleRequest_Should_ReturnFailureProgressEvent_When_DestinationIsFound_and_DestinationPolicyNotProvided() {
+        final DescribeDestinationsResponse describeResponse = DescribeDestinationsResponse.builder()
```
[the remark this answers is not in the document]
Comment (waifa) [reply aws-logs-destination/src/test/java/software/amazon/logs/destination/CreateHandlerTest.java:410]: I agree, was trying to be comprehensive and do most of the original test cases with the now optional destination policy removed. I see how that is redundant.

**Review by wchunl — APPROVED** [review]

File: aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java [hunk 8]
```diff
@@ -182,4 +182,83 @@ public void handleRequest_Should_ReturnSuccess_When_DestinationIsFound() {
...
+        assertThat(response.getStatus()).isEqualTo(OperationStatus.SUCCESS);
+        assertThat(response.getCallbackDelaySeconds()).isEqualTo(0);
+        assertThat(response.getResourceModel()).isEqualTo(request.getDesiredResourceState());
+        assertThat(response.getResourceModels()).isNull();
+        assertThat(response.getMessage()).isNull();
+        assertThat(response.getErrorCode()).isNull();
+    }
+
+    public void handleRequest_Should_ReturnSuccess_When_DestinationIsFound_and_DestinationPolicyNotProvided_PolicyExistsInOriginalDestination() {
+        final Destination destinationWithoutPolicy = getTestDestination(true);
```
Comment (evanuk) [inline aws-logs-destination/src/test/java/software/amazon/logs/destination/UpdateHandlerTest.java:226]: Should be `destinationWithPolicy`?

**Review by evanuk — APPROVED** [review]

**Review by wchunl — APPROVED** [review]

**Review by omkhegde — APPROVED** [review]
